### ISSUE TITLE
[binding events] shiny:inputchanged: regression or changes of guidelines ? Works in 1.1. Doesn't work in shiny 1.4

### ISSUE THREAD

**Issue by philibe** [rstudio/shiny#2676]
# Introduction of the issue : binding events

- My events example pasted in firefox debug console work with shiny 1.1 and shiny 1.4.
- The same events are  binded by `$(document).on('shiny:inputchanged',..)` with shiny_1.1.0
- The same events are __not__ binded by `$(document).on('shiny:inputchanged',..)` with shiny_1.4.0

I write the issue on my Windows computer at home with R 3.4, but I've discovered the initial issue on Linux Ubuntu at work with R 3.6.

# Test

## test of current reprex 

Go to every column search input and focus : the width of the current input search will grow (for shiny 1.1 and not for 1.4).

_Please modify `largeur_saisie_min` (=input width min) if your screen is too large to see the grow. Or add many columns like a real case._

## 2nd test not in this reprex but in the link which I talk below

Go to every __integer or numeric__ column search input and focus and type `$1` it will be replaced by `1...1` in the current input (for shiny 1.1 and not for 1.4)

# Expectation

To be able to bind all similar input of the __DOM document__.

# Context of recent update about `shiny:inputchanged`

I've read #1596 and #2442, #2446 the changes about `shiny:inputchanged` in News:  
> The shiny:inputchanged JavaScript event now triggers on the related input element instead of document. __Existing event listeners bound to document will still detect the event due to event bubbling__.

Except If I didn't understand the second sentence, binding events doesn't work anymore with `DOM document` (in my example of similar input in all of my application). 

Or the guidelines about binding events have changed even for `DOM document`.

_edit : (I have an 2nd example of event, [there in DT](https://github.com/rstudio/DT/issues/569#issuecomment-471767523) not written here,  and it's the same issue : it works with shiny 1.1 and not not with shiny 1.4. I don't think DT is the cause but probably shiny. In fact the more important issue for me it's this second linked example)._  

# Reprex

```R
library(shiny)
library(DT)
library(shinyjs)
app<-
  shinyApp(
    ui = basicPage(
      useShinyjs(),
      tags$head(
        HTML(
          "
          <script>
          $(document).on('shiny:inputchanged', function(event) {
            $('.datatables thead  input.form-control').on('focus',function(e){
              //console.log('focus');
              largeur_saisie_min=300;
              if (parseInt($(this).css('width')) <= largeur_saisie_min){
                $(this).css('width', largeur_saisie_min+'px');
              }
            });
          
            $('.datatables thead  input.form-control').on('focusout',function(e){
              //console.log('unfocus');
              $(this).css('width','100%');
            });
          });
          </script>
          "
        )
      ),
      fluidRow(column (12, DT::dataTableOutput('out1')) ),
      fluidRow(column (12, DT::dataTableOutput('out2')) )
    ),
    
    server = function(input, output) {
      
      df1 <-reactive(data.frame(num=c(1:2),char=LETTERS[1:2]))
      df2<-reactive(data.frame(num=c(3:4),char=LETTERS[3:4]))
      
      output$out1<- DT::renderDataTable(
        DT::datatable(df1(),  style = "bootstrap",   class = "compact", filter='top')
      )
      output$out2<- DT::renderDataTable(
        DT::datatable(df2(),  style = "bootstrap",   class = "compact", filter='top')
      )
    }
  )
shiny::runApp(app)
```
# The sessionInfo before and after
## sessionInfo: The event is binded by $(document).on('shiny:inputchanged',..) with shiny_1.1.0

```
sessionInfo()
R version 3.4.4 (2018-03-15)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
  [1] LC_COLLATE=French_France.1252  LC_CTYPE=French_France.1252    LC_MONETARY=French_France.1252 LC_NUMERIC=C                  
[5] LC_TIME=French_France.1252    

attached base packages:
  [1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
  [1] shinyjs_1.0 DT_0.9      shiny_1.1.0

loaded via a namespace (and not attached):
  [1] Rcpp_1.0.1        digest_0.6.18     later_0.8.0       mime_0.6          R6_2.4.0          jsonlite_1.6      xtable_1.8-4     
[8] magrittr_1.5      rlang_0.4.0       promises_1.0.1    tools_3.4.4       htmlwidgets_1.5.1 crosstalk_1.0.0   httpuv_1.5.2     
[15] yaml_2.2.0        compiler_3.4.4    htmltools_0.3.6  

```

## sessionInfo : The event is NOT binded by $(document).on('shiny:inputchanged',..) with shiny_1.4.0

```
sessionInfo()
R version 3.4.4 (2018-03-15)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
  [1] LC_COLLATE=French_France.1252  LC_CTYPE=French_France.1252    LC_MONETARY=French_France.1252 LC_NUMERIC=C                  
[5] LC_TIME=French_France.1252    

attached base packages:
  [1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
  [1] shinyjs_1.0 DT_0.9      shiny_1.4.0

loaded via a namespace (and not attached):
  [1] Rcpp_1.0.1        digest_0.6.18     later_0.8.0       mime_0.6          R6_2.4.0          jsonlite_1.6      xtable_1.8-4     
[8] magrittr_1.5      rlang_0.4.0       promises_1.1.0    tools_3.4.4       htmlwidgets_1.5.1 crosstalk_1.0.0   yaml_2.2.0       
[15] httpuv_1.5.2      fastmap_1.0.1     compiler_3.4.4    htmltools_0.4.0  
```

**Comment by nteetor** [comment]
Hello @philibe, I initiated this change. The input change event is now fired on the input getting changed. 

If you are running into problems with event bubbling, have you tried, 

```R
$(document).on("shiny:inputchanged",  ".datatables", function(event) {

})
```

If the `datatbles` class is not the top-level input class, please swap in the proper one.

**Comment by nteetor** [comment]
Followup, what browser are you using?

**Comment by nteetor** [comment]
There is also a chance the problem is related to [this issue](https://github.com/rstudio/htmltools/issues/143), can you open your browser's console to see if there are any errors?

**Comment by philibe** [comment]
Hello,

TLDR: I Would like an equivalent of `$('document').ready(function(){});` for shiny. :) Reading many articles it   seemed that was `$(document).on('shiny:inputchanged'`.

>I initiated this change. 

I saw that :)

> The input change event is now fired on the input getting changed.

It's a good thing.... **only if there is not regression or an another guidance**.  :)

>`$(document).on('shiny:inputchanged',  ".datatables"`

I've tried your suggestion and it doesn't work. And it seems to me logical : 

**I would like to bind these events when shiny have loaded all objects, not before, not after.** :)

- **not before, too early** : Shiny by Ajax have not yet build the components. The initial event on DOM Document (`$(document).on('shiny:inputchanged'`) was for test that, like many examples I've read on github, Stack Overflow, Rstudio, because with `$('document').ready(function(){});` it's too early in Shiny Ajax application.
- **not after, too late** : Your suggestion is to wait to focus on the `.datatables`, I've tried also a top in put class like `.container-fluid`, but it's too late to built the events  **I would like the event built at the beginning** , like `$('document').ready(function(){});` in straight html application without Ajax to build.  

>Followup, what browser are you using?
At home : Firefox 69.0.1, and sometimes (Google Chrome Version 77.0.3865.120, Microsoft Edge 44.18362.387.0)

**Comment by nteetor** [comment]
Some clarifying questions,

1. Are you hoping to run javascript code every time a table changes?
1. Are you hoping to run javascript code once (and only once) when the application is initialized?
1. How familiar are you with javascript event delegation?

**Comment by philibe** [comment]
1# no
2# once
    once to bind the event at the beginning
    at each time I focus the input, the input will grow, and at each time I focus out, it will shrink.
    
3# in general yes, but:

I know that in DT

- https://shiny.rstudio.com/reference/shiny/latest/renderDataTable.html

I didn't know *Shiny* that:

- https://shiny.rstudio.com/articles/js-custom-input.html


If I have many input not only in DT, why to use complex callback: before this change everything worked.

I would like only `('document').ready(function(){}); `

But you have my reprex, could you demonstrate me the usage: it's feel very heavy to implement my little code.

**Comment by nteetor** [comment]
I phrased my questions poorly. However, you would like some javascript code to run potentially multiple times during the lifetime of your application. I think I am on the same page now. 

Event delegation removes the problem of attaching events to elements that do not yet exist. Fortunately, the `document` object is present once the dom has loaded, so we are safe to attach delegated event handlers very quickly on startup. Take a look at the line below,

`$(document).on("click", ".my-button", function(event) { . . .`

This line says we are attaching an event handler to the `document` object. The handler will trigger whenever the _child element with class `my-button`_ is clicked. Note, this button does not have to exist when the javascript is initially run. Also, any valid CSS selector may be used instead of `".my-button"`.

You will want to drop such handlers into either of the blocks below, [they are equivalent](https://learn.jquery.com/using-jquery-core/document-ready/). 

```javascript
$(function() {
  // assign handlers here
});
```

```javascript
$(document).ready(function() {
  // assign handlers here
});
```

Finally, the focus event problem is a separate problem and outside the scope of this repo. The problem has to do with manually focusing an element versus an element automatically gaining focused when the user interacts with it. The latter does not appear to happen. In the example below user interactions with the table have no effect on the focus state. Clicking one of the buttons will manually force focus onto the table and the corresponding event handler will show an alert. In general, I believe the focus and focusin events are reserved for input elements, https://developer.mozilla.org/en-US/docs/Web/API/Element/focusin_event. I would encourage you to continue the conversation over at https://community.rstudio.com. 

```R
library(shiny)

js <- HTML("
$(function() {
  $(document).on('focusin', '.datatables', function(event) {
    // This never runs!
    alert('Table ' + event.currentTarget.id + ' was focused');
  });
  
  $(document).on('click', '.my-button', function(event) {
    alert('Clicked button with class my-button');
  })
});
");

shinyApp(
  ui = fluidPage(
    tags$head(tags$script(js)),
    fluidRow(
      column(
        width = 6,
        DT::dataTableOutput("table_cars")
      ),
      column(
        width = 6,
        tags$div(
          class = "my-div",
          "I'm another div"
        )
      )
    ),
    tags$button("Force focus on table", onclick = "$(document.querySelector('.datatables')).focusin()"),
    tags$button(class = "my-button", "I show an alert")
  ),
  server = function(input, output) {
    output$table_cars <- DT::renderDataTable({
      DT::datatable(mtcars)
    })
  }
)

```

**Comment by philibe** [comment]
Thank you for your detailed explanations, but I would like only my little reprex modified with your point of vue, and only after I can believe your point of view in my case and the theories.

But thank you, I modified your examples, and I get what I expected. :)

Is it in a good way ?

```R
library(shiny)

js <- HTML("
 function search_replace_interval(event) {
    //console.log(event.data.column_type);

    if (event.keyCode == 13) {
        before = $(this).val();
        after = before;
        // prefix by the $ if you are looking for a number without interval : 
        //$34 will be replaced by 34...34 when type of column is integer (or date)
        //$34 will be replaced by 33...35 when type of column is number (because the previous case is misplaced when number)
        //$$34 will be replaced by 33...35 (to workaround the previous case when misplaced)
        item = before.substring(before.lastIndexOf('$') + 1);
        if (before.substring(0, 1) == '$') {
            if (((before.substring(0, 2) == '$$') || (event.data.column_type == 'number')) &&
                (['number', 'integer'].indexOf(event.data.column_type) >= 0)) {
                after = eval(eval(item) - 1) + '...' + eval(eval(item) + 1);
            } else {
                after = item + '...' + item;
            }
            //console.log(after);
        }
        $(this).val(after)
    }
};
$(function() {
    $(document).on('focusin', '.datatables', function(event) {
        $('.datatables thead  input.form-control').on('focus', function(e) {
            //console.log('focus');
            largeur_saisie_min = 300;
            if (parseInt($(this).css('width')) <= largeur_saisie_min) {
                $(this).css('width', largeur_saisie_min + 'px');
            }
        });

        $('.datatables thead  input.form-control').on('focusout', function(e) {
            //console.log('unfocus');
            $(this).css('width', '100%');
        });
        $('.datatables thead [data-type~=\"number\"] input.form-control').on('keydown', {
            column_type: 'number'
        }, search_replace_interval);
        $('.datatables thead [data-type~=\"integer\"] input.form-control').on('keydown', {
            column_type: 'integer'
        }, search_replace_interval);
        $('.datatables thead [data-type~=\"date\"] input.form-control').on('keydown', {
            column_type: 'date'
        }, search_replace_interval);
        $('.datatables thead [data-type~=\"time\"] input.form-control').on('keydown', {
            column_type: 'time'
        }, search_replace_interval);

    });

    $(document).on('click', '.my-button', function(event) {
        alert('Clicked button with class my-button');
    })
});
           ");

shinyApp(
  ui = fluidPage(
    tags$head(tags$script(js)),
    fluidRow(
      column(
        width = 6,
        DT::dataTableOutput("table_cars")
      ),
      column(
        width = 6,
        tags$div(
          class = "my-div",
          "I'm another div"
        )
      )
    ),
    tags$button("Force focus on table", onclick = "$(document.querySelector('.datatables')).focusin()"),
    tags$button(class = "my-button", "I show an alert")
  ),
  server = function(input, output) {
    output$table_cars <- DT::renderDataTable({
      DT::datatable(mtcars,  style = "bootstrap",   class = "compact", filter='top')
    })
  }
)
```

**Comment by philibe** [comment]
So, If I understand :

#2442 was committed to avoid that "the shiny:inputchanged event is not fired on a changing input element and instead is fired on document."

But it avoid also a sort of document ready, so it's a sort of regression, but it was not a good practice and the most important: the other manner to catch events work with shiny and is a good practice. 

So I close :). The exemple I've seen was not all correct with  a sort of document ready but I thinked that other manners don't work with the Ajax of Shiny.

Only a request : the Shiny documentation is not clear for that because there is a long page to explain complex 
events but not for simple use. It's a little the cause of my issue :P


Maybe #2442 will fix my issue  in my[addendum 2 comment in ly SO "Get the event which is fired in Shiny and with grep (generated input)" ](https://stackoverflow.com/questions/56776207/get-the-event-which-is-fired-in-shiny-and-with-grep-generated-input/56778086?noredirect=1#comment100169780_56778086) when I read "the shiny:inputchanged event is not fired on a changing input element and instead is fired on document.".

**Comment by philibe** [comment]
@nteetor : a little question.

Your solution works for all events except for the `load` events.

For example this example below doesn't work if I replace `focusin` by `load`, ie when I want add/modify an element only when all shiny elements are drawn.


```javascript
$(document).on('focusin', '.datatables', function(event) {
    $(this).find('table.dataTable').closest('div').addClass('div_containing_table');
    console.log($(this).find('table.dataTable'));
    console.log('coucou 2');
}); 

```

Please, How can I do that  ? :) I've seen [jquery.initialize on SO](https://stackoverflow.com/questions/15268661/jquery-on-create-event-for-dynamically-created-elements/49023264#49023264), but I hope there is Shiny solution.

If I can't is there  a lack in the solution ? Note that is not a challenge or support out of this issue. It's a real question of mine, today about #1596 and #2442, #2446. In other terms to reopen this issue :)


Knowing that from [SO](https://stackoverflow.com/questions/11885425/onload-of-new-element-delegate-on/11885723#11885723) redirecting from [jquery additionnal notes](https://api.jquery.com/on/#additional-notes)
>"In all browsers, the load, scroll, and error events (e.g., on an element) do not bubble. [...] Such events are not supported for use with delegation, but they can be used when the event handler is directly attached to the element generating the event."

**Comment by philibe** [comment]
I reopen for the `load` event as explained in last message, because for this particular event there is a regression. 

Tell me if I have to create a "feature requested", or a new issue, if you think it's better :)
